---
title: "RA Ex Vivo 16S Analysis"
author: "Renuka Nayak"
date: "8/20/20"
output: 
  html_document:
    toc: true 
    toc_float: true
    number_sections: true
---

# Introduction

This R markdown document provides code used in the analysis of 16S-seq data from RA patient fecal samples treated ex vivo with either MTX or vehicle control (DMSO). These are from our manuscript:

Nayak et al, Methotrexate impacts conserved pathways in diverse human gut bacteria leading to decreased host immune activation. In revision.  

#  Experimental Design

To test the impact of MTX in the context of a complex human gut microbial community we incubated stool samples obtained from 30 MTX-na√Øve patients (Table S1A) with MTX (100 ug/ml) or vehicle for 48 hours. Quadruplicate samples before and 24 hours after treatment from 4 of these individuals were subjected to 16S sequencing.

## Analysis Notes

Some of the code in this document has been adapted from: https://f1000researchdata.s3.amazonaws.com/manuscripts/10545/387d9a79-c4ab-4676-8466-583e3fb2c85f_8986_-_susan_holmes_v2.pdf?doi=10.12688/f1000research.8986.2

## Manuscript Results

The code in this Rmd was used to support the following paragraph from the Results of our manuscript:

"16S-seq of endpoint samples revealed a significant decrease in richness (Figure S4C) and shift in community structure (Figures 5B-E) in response to MTX. Consistent with our studies in mice and on isolates, Bacteroidetes decreased (Figure 5F) and Actinobacteria increased (Figure 5G). We also detected significant shifts in the abundance of 17 genera and 20 ASVs (Table S2C, Figure 5H). This included multiple MTX-modulated ASVs seen in our in vivo experiments: B. thetaiotaomicron (ASV23), B. uniformis (ASV14), B. vulgatus (ASV26), C. aerofaciens (ASV44), and B. ovatus (ASV496), which changed in a direction similar to what was seen in vivo.""

# Set Up

## Load Libraries

```{r setup, include=TRUE}
knitr::opts_chunk$set(echo = TRUE, message = FALSE, warning = FALSE)
library(phyloseq); packageVersion("phyloseq")
library(ggplot2); packageVersion("ggplot2")

library(biomformat)
library(ape)
library(readxl)
library(ggthemes)
library(tidyverse)
library(RColorBrewer)
library(gridExtra)
library(qiime2R)
library(nlme)

# Load packages into session, and print package version
#sapply(c(.cran_packages, .bioc_packages), require, character.only = TRUE)
set.seed(100)

theme_set(theme_classic(base_size = 9) +theme(axis.text = element_text(colour = "black"), legend.position="right", strip.background = element_blank()))
```

## Set Path Names
Set path names where the data and results will be found (variables path and f):  
`path=<parent directory where QIIME2 pipeline results are found>`
`sampleDF_RA_fn=<excel file with info on samples>`
```{r, include=FALSE}
path="/Volumes/turnbaughlab/qb3share/rnayak/13_gnoto9RA_16S/04_qiime2Analysis/Output/"
home="renukanayak"
work="rnayak"
site=work
sampleDF_gnoto9_fn=paste0("/Users/",site,"/Box/Renuka1/MouseExperiments/20190311_gnoto9/20190311_gnoto9data.xlsx")
sampleDF_RA_fn=paste0("/Users/",site,"/Box/Renuka1/MouseExperiments/20190612_16SGnoto789/05_exVivoSampleInfo/20190716_RAexvivoSampleSheet.xlsx")
```

## Obtain Data from the QIIME2 Pipeline
```{r message=FALSE, warning=FALSE}
SVtable=read_qza(paste0(path,"SVtable.qza"))
SVseq=read_qza(paste0(path,"SVsequences.qza"))
SVtree=read_qza(paste0(path,"SVtree_denovo.qza"))
taxonomy=read_tsv(paste0(path,"SV_taxonomy_Dada2.txt"))
#add md5 as rownames
taxonomy=taxonomy %>% left_join(as.data.frame(SVseq$data) %>%
                                  rownames_to_column("md5"),
                                by=c("SV"="x"))
rownames(taxonomy)=taxonomy$md5
taxonomy=as.matrix(taxonomy)[,c(2:8,1)] #rearrange to help with tax_glom
```

## Sample Information

These samples were sequenced along with samples from a gnotobiotic mouse transplant experiment. Information for both types of samples are used below. 

```{r}

#FMT samples
sampleDF_gnoto9=read_xlsx(sampleDF_gnoto9_fn, sheet="11_16S")
sampleDF_gnoto9$`Sample Label`=str_c("RRN_",sampleDF_gnoto9$`Sample Label`)
sampleDF_gnoto9_org=sampleDF_gnoto9 %>% select("Sample Label", "mouseID", 
                                               "Feces Time Point", "Day", 
                                               "Tx","Cage","Isolator",
                                               "SampleInfo","Row","Column","type")
rownames(sampleDF_gnoto9_org)=sampleDF_gnoto9_org$`Sample Label`
colnames(sampleDF_gnoto9_org)=c("Sample Label", "MouseID", "TimePoint", 
                                "Day", "Tx","Cage","Isolator",
                                "SampleInfo","Row","Column","type")
sampleDF_gnoto9_org$exp="gnoto9"

sampleDF_RA=read_xlsx(sampleDF_RA_fn, sheet="01_16S")

#RA ex vivo samples
sampleDF_RA_org=sampleDF_RA %>% select("Sample Label", "Patient(baselineSample)", 
                                       "TimePoint(hrs)", "Treatment",
                                       "SampleInfo","Row","Column", "type")
rownames(sampleDF_RA_org)=sampleDF_RA_org$`Sample Label`
colnames(sampleDF_RA_org)=c("Sample Label", "Patient", "TimePoint", "Tx",
                            "SampleInfo","Row","Column","type")
sampleDF_RA_org$exp="RAexVivo"
sampleDF_RA_org$TimePoint=as.character(sampleDF_RA_org$TimePoint)

sampleDF=full_join(sampleDF_gnoto9_org, sampleDF_RA_org)
rownames(sampleDF)=sampleDF$`Sample Label`
```

## Create Phyloseq Object

```{r}
physeq<-phyloseq(
  otu_table(SVtable$data, taxa_are_rows = T),
  phy_tree(SVtree$data),
    tax_table(taxonomy),
  sample_data(sampleDF %>% as.data.frame()))
```

```{r}
ps <- physeq 
#filter taxa based on the number of counts found in that taxa
ps = filter_taxa(ps, function(x) sum(x) > 10, TRUE)
ps
rank_names(ps)
table(tax_table(ps)[, "Phylum"], exclude = NULL)
#remove taxa that are uncharacterized at the phylum level
ps0 <- subset_taxa(ps, !is.na(Phylum) & !Phylum %in% c("", "uncharacterized")  & !Kingdom %in% c("Eukaryota"))
ps0
```


## Examine Phylums Represented
The vast majority of ASVs are Firmicutes.
```{r}
#From the phyloseq workflow on F1000
# Compute prevalence of each feature, store as data.frame
prevdf = apply(X = otu_table(ps0),
               MARGIN = ifelse(taxa_are_rows(ps0), yes = 1, no = 2),
               FUN = function(x){sum(x > 0)})

# Add taxonomy and total read counts to this data.frame
prevdf = data.frame(Prevalence = prevdf,
                    TotalAbundance = taxa_sums(ps0),
                    tax_table(ps0))

plyr::ddply(prevdf, "Phylum", function(df1){cbind(mean(df1$Prevalence),sum(df1$Prevalence))})

# Subset to the remaining phyla
prevdf1 = subset(prevdf, Phylum %in% get_taxa_unique(ps0, "Phylum"))

ggplot(prevdf1, aes(TotalAbundance, Prevalence / nsamples(ps0),color=Phylum)) +
  # Include a guess for parameter
  geom_hline(yintercept = 0.05, alpha = 0.5, linetype = 2) + geom_point(size = 2, alpha = 0.7) +
  scale_x_log10() + xlab("Total Abundance") + ylab("Prevalence [Frac. Samples]") +
  facet_wrap(~Phylum) + theme(legend.position="none")

# Define prevalence threshold as 5% of total samples
prevalenceThreshold = 0.05 * nsamples(ps0)
prevalenceThreshold

#Report number in each taxanomic rank
for (i in rank_names(ps)) {
  print(paste("Number of ",i,":",sep=""))
  print(length(get_taxa_unique(ps0, taxonomic.rank = i)))
}
```
## Remove Samples with Low Reads
```{r}
#remove samples with fewer than 1000 reads
arbitraryCutOff=1000
ps0 <- prune_samples(sample_sums(ps0)>=arbitraryCutOff, ps0)
```

## Transform Data
Create transformed versions of the data to perform subsequent analyses.

```{r}

#Convenience functions for geometric mean and clr transform; all x are > 0 because of filtering that was done
gm_mean = function(x, na.rm=TRUE){
  # The geometric mean, with some error-protection bits.
  exp(sum(log(x[x > 0 & !is.na(x)]), na.rm=na.rm) / length(x))
  }

clr = function(x, base=2){
  x <- log((x / gm_mean(x)), base)
  x[!is.finite(x) | is.na(x)] <- 0.0
  return(x)
  }

#clr transform data for PCoA, multivariate analysis
ps_clr <- transform_sample_counts(ps0, clr )

#proportion data for visualization
ps_prop <- transform_sample_counts(ps0, function (x) x/sum(x))

#rarefied data for ANOSIM and PERMANOVA
set.seed(25123)
ps_rarefied <- rarefy_even_depth(ps0, sample.size=1000) 
```
 

# Analysis of RA Ex Vivo Samples

## Alpha Diversity
```{r}

ps_RA=subset_samples(ps0, exp=="RAexVivo")
a=sample_data(ps_RA)
a$well=str_c(a$Row,a$Patient)
sample_data(ps_RA) <- a
a <- as.data.frame(sample_data(ps_RA))

er_ra=estimate_richness(subset_samples(ps_RA, TimePoint=="24" & SampleInfo=="ex vivo"))
er_ra2=rownames_to_column(er_ra, "Sample.Label") %>% left_join(as_tibble(a), by="Sample.Label")
er_ra2$Patient=factor(er_ra2$Patient, levels=c("215","347", "72","211"))
er_ra2$Tx=as.factor(er_ra2$Tx)
ggplot(er_ra2, aes (x=Tx, y=Observed, group=Patient, fill=Tx)) + geom_bar(stat="summary", fun.y="mean") + 
  geom_point(shape=21) + facet_wrap(~Patient, nrow=1, scales="free") + theme(legend.position = "none")

library(nlme)
er_nlme=lme(Observed~Tx, random=~1|Patient, data=er_ra2, na.action=na.omit)
summary(er_nlme)
```

There is a reduction in alpha diversity (p=3x10-4, linear mixed effects model).

## Beta Diversity
Examine PCA plots using clr transformed data.

```{r}
ps_t <- ps_clr
ps_t_RA <- subset_samples(ps_t, exp=="RAexVivo" & SampleInfo=="ex vivo") 
plotList=list()
patients=c(215,347, 72,211)
for (i in 1:4) {
   
   ps_t_pt <- subset_samples(ps_t_RA, Patient==patients[i] & TimePoint %in% c(0,24))
   PS.ord <- ordinate(ps_t_pt, "RDA", "euclidean")
  
  ggP_df<- plot_ordination(ps_t_pt, PS.ord, type="samples", color="Tx", justDF = T)
  ggP <- ggplot(data=ggP_df,aes(x=PC1,y=PC2,  fill=Tx, label=TimePoint)) + geom_point(shape=21) +
    facet_grid(TimePoint~.) +
    theme (panel.border = element_rect(color = "black", fill = NA, size = 1),
           strip.background = element_rect(color = "black", size = 1)) +
    ggtitle(paste("Patient",patients[i]))
  print(ggP)

  plotList[[i]] <- ggP
}
```

## Hypothesis Testing

### ANOSIM
```{r}
set.seed(25123)
ps_rarefied <- rarefy_even_depth(ps_RA, sample.size=5000)
library("vegan")
packageVersion("vegan")
set.seed(10)
ps_t <- ps_rarefied

#Difference at 0 hours
for (i in c(0)) {
  for (j in c(215, 347,72, 211)) {
    ps_ano <- subset_samples(ps_t, TimePoint %in% c(i) & Patient==j & SampleInfo=="ex vivo")
    mtx_group = get_variable(ps_ano, "Tx")
    mtx_ano = anosim(phyloseq::distance(ps_ano, "bray"), mtx_group)
    print(mtx_ano)
  }
}

#Difference at 24 hours
for (i in c(24)) {
  for (j in c(215, 347,72, 211)) {
    ps_ano <- subset_samples(ps_t, TimePoint %in% c(i) & Patient==j & SampleInfo=="ex vivo")
    mtx_group = get_variable(ps_ano, "Tx")
    mtx_ano = anosim(phyloseq::distance(ps_ano, "bray"), mtx_group)
    print(mtx_ano)
  }
}
```

### PERMANOVA
```{r}
#PERMANOVA testing
# Adonis test to determine whether methotrexate tx influences community composition at 24h.
set.seed(1)
ps_t <- ps_rarefied
ps_perma <-  subset_samples(ps_t, TimePoint %in% c(24) & SampleInfo=="ex vivo")
ps_perma_bray <- phyloseq::distance(ps_perma, method = "bray")
sampledf <- data.frame(sample_data(ps_perma))
adonis(ps_perma_bray ~ Tx + Patient, data = sampledf)

print("beta dispersion testing")
beta <- betadisper(ps_perma_bray, sampledf$Tx)
permutest(beta) #tests whether differences are due to group dispersions (if they are, then p will be significant)
```

Suggests that at 24 hours there is a difference between MTX and DMSO.

## Phylum Level Trends

Perform stats on phylum level changes by patient for other phyla
```{r}

ps_phylum0 = tax_glom(ps_RA, "Phylum", NArm = TRUE) 
ps_phylum = transform_sample_counts(ps_phylum0, clr) 

ps_p <- subset_samples(ps_phylum, TimePoint %in% c(24) & SampleInfo=="ex vivo") #PICK A TIMEPOINT to examine

for (i in c(2,3,5,6,7)) { 
  #1 and 4 are excluded because they are  Euryarchaeota and Tenericutes, 
  #which throw an error (overfit models) lb/c of low abundance/prevalence
  
  #visualize
  ph <- rownames(tax_table(ps_p))
  o <- ph[i]
  o_names <- as(tax_table(ps_p)[o,], "matrix")
  o_title <- paste(o_names[1],o_names[2], sep="_")
  y_ps = otu_table(ps_p)[o,]
  y_df = psmelt(y_ps)
  y_df <- cbind(y_df, sample_data(ps_p)[y_df$Sample, c("Tx", "Patient","well")])
  colnames(y_df) = c("OTU","Sample","Abundance","Tx", "Patient","well")
  y_df$Patient=factor(y_df$Patient, levels=c("215","347","72", "211"))
  g <- ggplot(y_df) +  aes(x=Tx, y=Abundance, fill=Tx) +  
    geom_bar(stat = "summary", fun.y = "mean") + 
    geom_point(shape=21, color="black") + 
    xlab("Treatments") + facet_grid(~Patient) + 
    theme(axis.text.x = element_text(angle = 90, hjust =1, vjust=0.5), legend.position = "none") +
    ggtitle(o_title)
  print(g)
 
  #mixed effects linear model to test effect of tx on abundace 
  b_nlme <- lme(Abundance~Tx, random= ~1|Patient, data= y_df)
  print(o_names[2])
  print(summary(b_nlme)$tTable[,5][2])
}

```

# DESeq Analysis

## ASV Level
```{r}

library(DESeq2)
#Use Deseq to identify differentially abundant RSVs
#https://www.bioconductor.org/packages/devel/bioc/vignettes/phyloseq/inst/doc/phyloseq-mixture-models.html

alpha = 0.01 #SET significance threshold use 0.01 for RSVs throughout the manuscript
set.seed(10)

#time 24h 
ps_deseq <- subset_samples(ps_RA,SampleInfo=="ex vivo")
ps_deseq=subset_samples(ps_deseq, 
                        Tx %in% c("DMSO","MTX") & TimePoint %in% c(24)) 
deseq_mtx=phyloseq_to_deseq2(ps_deseq, ~ Patient + Tx)
geoMeans = apply(counts(deseq_mtx), 1, gm_mean)
deseq_mtx = estimateSizeFactors(deseq_mtx, geoMeans = geoMeans)
deseq_mtx = DESeq(deseq_mtx, test="Wald", fitType="parametric")
res = results(deseq_mtx, cooksCutoff = FALSE)
res2 = cbind(as(res, "data.frame"), as(tax_table(ps_deseq)[rownames(res), ], "matrix"))
sigtab = res[which(res$padj < alpha), ]  #SET significance threshold above
sigtab = cbind(as(sigtab, "data.frame"), as(tax_table(ps_deseq)[rownames(sigtab), ], "matrix"))
dim(sigtab)
print(paste("There are", dim(sigtab)[1],"differentially abundant ASVs with MTX treatment at 24h."))
sigtab_24 <- sigtab
res2_24 <- res2
```


### Phylogenetic Tree

Tree of the ~20 RSVs that were differentially abundant at time 24
```{r}
library(ggtree)
library(tidytree)
library(plotly)

ra_ps=subset_samples(ps_RA, SampleInfo=="ex vivo")
ra_ps=filter_taxa(ra_ps, function(x) sum(x) > 0, TRUE)
ra_ps_sig=prune_taxa(rownames(sigtab_24), ra_ps)
ra_tree=phy_tree(ra_ps_sig)
rtt=as_tibble(ra_tree)
rtt2=left_join(rtt,rownames_to_column(sigtab_24,"label"),by="label")
ra_tree2=as.treedata(rtt2)

#visualize with a number of parameters
p <- ggtree(ra_tree2,layout="circular") +
  geom_point(aes(size=abs(log2FoldChange),color=as.factor(sign(log2FoldChange)), shape=Phylum))  

#rotate branch to make proteobacteria together
ggtree(ra_tree2,layout="circular") + geom_text(aes(label=node))
flip(p, 35, 37)

#visualize again
p <- ggtree(ra_tree2,layout="circular") +
  geom_point(aes(size=abs(log2FoldChange),color=as.factor(sign(log2FoldChange)), alpha=0.05))
flip(p, 35, 37)
```

### Genus Level DESeq
```{r}
library(DESeq2)
#Use Deseq to identify differentially abundant RSVs
#https://www.bioconductor.org/packages/devel/bioc/vignettes/phyloseq/inst/doc/phyloseq-mixture-models.html

alpha = 0.01 #SET significance threshold use 0.01 for RSVs throughout the manuscript
set.seed(10)
ps_RAgenus=tax_glom(ps_RA, "Genus", NArm=T)

#time 24h 
ps_deseq <- subset_samples(ps_RAgenus,SampleInfo=="ex vivo")
ps_deseq=subset_samples(ps_deseq, 
                        Tx %in% c("DMSO","MTX") & TimePoint %in% c(24)) 
deseq_mtx=phyloseq_to_deseq2(ps_deseq, ~ Patient + Tx)
geoMeans = apply(counts(deseq_mtx), 1, gm_mean)
deseq_mtx = estimateSizeFactors(deseq_mtx, geoMeans = geoMeans)
deseq_mtx = DESeq(deseq_mtx, test="Wald", fitType="parametric")
res = results(deseq_mtx, cooksCutoff = FALSE)
res2 = cbind(as(res, "data.frame"), as(tax_table(ps_deseq)[rownames(res), ], "matrix"))
sigtab = res[which(res$padj < alpha), ]  #SET significance threshold above
sigtab = cbind(as(sigtab, "data.frame"), as(tax_table(ps_deseq)[rownames(sigtab), ], "matrix"))
dim(sigtab)
print(paste("There are", dim(sigtab)[1],"differentially abundant genera with MTX treatment at 24h."))
sigtab_24_genus <- sigtab # 16 significant
res2_24 <- res2
```
16 genera differentially abundant at time 24h and none at 0 hour. 


## Phylum Level DESeq
```{r}
library(DESeq2)
#Use Deseq to identify differentially abundant RSVs
#https://www.bioconductor.org/packages/devel/bioc/vignettes/phyloseq/inst/doc/phyloseq-mixture-models.html

alpha = 0.01 #SET significance threshold use 0.01 for RSVs throughout the manuscript
set.seed(10)
ps_RAphylum=tax_glom(ps_RA, "Phylum", NArm=T)

#time 24h 
ps_deseq <- subset_samples(ps_RAphylum,SampleInfo=="ex vivo")
ps_deseq=subset_samples(ps_deseq, 
                        Tx %in% c("DMSO","MTX") & TimePoint %in% c(24)) 
deseq_mtx=phyloseq_to_deseq2(ps_deseq, ~ Patient + Tx)
geoMeans = apply(counts(deseq_mtx), 1, gm_mean)
deseq_mtx = estimateSizeFactors(deseq_mtx, geoMeans = geoMeans)
deseq_mtx = DESeq(deseq_mtx, test="Wald", fitType="parametric")
res = results(deseq_mtx, cooksCutoff = FALSE)
res2 = cbind(as(res, "data.frame"), as(tax_table(ps_deseq)[rownames(res), ], "matrix"))
sigtab = res[which(res$padj < alpha), ]  #SET significance threshold above
sigtab = cbind(as(sigtab, "data.frame"), as(tax_table(ps_deseq)[rownames(sigtab), ], "matrix"))
dim(sigtab)
print(paste("There are", dim(sigtab)[1],"differentially abundant phyla with MTX treatment at 24h."))
sigtab_24_phy <- sigtab 
res2_24 <- res2
```
Bacteroidetes decrease and Actinobacteria increase at 24 hours.

#Session Info
```{r}
sessionInfo()

```

